packages/transaction-controller/src/gas-flows/SmartTransactionGasFeeFlow.ts

```
@@ -0,0 +1,132 @@
+import { hexToBN, toHex } from '@metamask/controller-utils';
+import { createModuleLogger, type Hex } from '@metamask/utils';
+import type BN from 'bn.js';
+
+import { projectLogger } from '../logger';
+import type {
+  GasFeeEstimates,
+  GasFeeFlow,
+  GasFeeFlowRequest,
+  GasFeeFlowResponse,
+  TransactionMeta,
+  getSmartTransactionFeeEstimatesResponse,
+} from '../types';
+import { GasFeeEstimateLevel } from '../types';
+import { pickMiddleFeeElement } from '../utils/utils';
+import { DefaultGasFeeFlow } from './DefaultGasFeeFlow';
+
+const log = createModuleLogger(projectLogger, 'smart-transaction-gas-fee-flow');
+
+const BASE_FEE_MULTIPLIERS = {
+  low: 1,
+  medium: 1.35,
+  high: 1.7,
+};
+
+const PRIORITY_FEE_MULTIPLIERS = {
+  low: 1,
+  medium: 1.05,
+  high: 1.1,
+};
+
+type FeesByLevel = {
+  [key in GasFeeEstimateLevel]: BN;
+};
+
+/**
+ * Implementation of a smart transaction gas fee flow specific to Smart Transactions.
+ */
+export class SmartTransactionGasFeeFlow implements GasFeeFlow {
+  matchesTransaction(transactionMeta: TransactionMeta): boolean {
+    return Boolean(transactionMeta.isSmartTransaction);
+  }
+
+  async getGasFees(request: GasFeeFlowRequest): Promise<GasFeeFlowResponse> {
+    try {
+      return await this.#getSmartTransactionGasFees(request);
+    } catch (error) {
+      log('Using default flow as fallback due to error', error);
+      return new DefaultGasFeeFlow().getGasFees(request);
```

Is this a reasonable and safe fallback for smart transaction gas estimation? Should we bubble the error up to the user. Just asking since I'm not familiar with the design choices behind this.

---

I was also a bit unsure that we need to use `DefaultGasFeeFlow` or not. But from what I understand Matthew's message here https://consensys.slack.com/archives/C03RV0Q4TGB/p1709203816115559?thread_ts=1709201748.892819&cid=C03RV0Q4TGB
any kind of transaction could be a smart one in future, hence defaulting here should be fine.